lightningd/pay.c

```diff
@@ -1070,3 +1071,103 @@ static const struct json_command listsendpays_command = {
 	"Show sendpay, old and current, optionally limiting to {bolt11} or {payment_hash}."
 };
 AUTODATA(json_command, &listsendpays_command);
+
+static struct command_result *json_createonion(struct command *cmd,
+						const char *buffer,
+						const jsmntok_t *obj UNNEEDED,
+						const jsmntok_t *params)
+{
+	const jsmntok_t *hopstok, *hop, *payloadtok, *typetok, *pubkeytok;
+	enum sphinx_payload_type type;
+	size_t i, hop_count = 0;
+	struct json_stream *response;
+	struct pubkey pubkey;
+	struct secret *session_key, *shared_secrets;
+	struct sphinx_path *sp;
+	u8 *assocdata, *payload, *serialized;
+	struct onionpacket *packet;
+
+	if (!param(cmd, buffer, params,
+		   p_req("hops", param_array, &hopstok),
+		   p_req("assocdata", param_bin_from_hex, &assocdata),
+		   p_opt("session_key", param_secret, &session_key),
+		   NULL)) {
+		return command_param_failed();
+	}
+
+	if (session_key == NULL)
+		sp = sphinx_path_new(cmd, assocdata);
+	else
+		sp = sphinx_path_new_with_key(cmd, assocdata, session_key);
+
+	json_for_each_arr(i, hop, hopstok) {
```

Comment:
It's better, BTW, to do the parsing in a new param_hops_array.  That way you get proper checking if you use the 'check' rpc command.  Might be too awkward though?